clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -4793,6 +4803,155 @@ void CodeGenModule::addGlobalIntelFPGAAnnotation(const VarDecl *VD,
   }
 }
 
+static Optional<std::string>
+getValidAttributeNameAsString(const Expr *NameE, const ASTContext &Context) {
```

Comment:
@AaronBallman/ @erichkeane / @premanandrao  could you please take a look at these functions generating valid name and value string for `add_ir_attributes_*` attribute. I am not very familiar with this.

Reply:
@AaronBallman | @erichkeane | @premanandrao - Friendly ping. It is not very elegant code, so I am very curious to hear if there are better ways to do it.

Reply:
Just poking in here: What is this trying to do?  Getting the name of the attribute shouldn't require this much work, and attributes have their values?  I've only looked at a little of this CodeGenModule.cpp part, and I suspect there is a need to re-imagine what this patch should look like.  There is a lot of work happening here I'm unconvinced is worth while.

Reply:
I can see how the name is a little misleading. This extracts the string value from an expression that was passed to one of the `[[__sycl_device__::add_ir_attributes_*(...)]]`. These expressions must either be a string literal or a `constexpr const char *`. Extracting the value of a string literal is pretty straight forward, but I could not find a better way of getting the string value from a `constexpr const char *`.

Similarly, `getValidAttributeValueAsString` also allows a handful other types in addition to strings.

Reply:
I fear it's only a very small part of this that could be moved to `APValue` and even then, since it recursively extracts from the underlying expression of `APValue` in the same way it would any other `Expr`, it may end up having more repeat code.